### ISSUE TITLE
wrong segment context during interrupt service routine

### ISSUE THREAD

**Issue by ladmanj** [tkchia/gcc-ia16#94]
Hi @tkchia and all

I have encountered another weird problem.
Yes probably because of my lack of platform knowledge again.

BEWARE: This hardware is not a PC!

I have couple of Interrupt Service routines written in C one is:

```
void __attribute__((interrupt)) __far  ISR_10_usart0_rx(void)
{
	buf_add_element(u0_rec_handle, io_read_8(USART0));
	if(buf_count_occupied(u0_rec_handle)>(BUFFER_SIZE/2))
	{
		usart_disable_flag(USART0, (UCMD_DTR | UCMD_RTS));
	}
	else
	{
		usart_enable_flag(USART0, UCMD_DTR | UCMD_RTS);
	}
}
```
The resulting code is:
```
000094b4 <ISR_10_usart0_rx>:
    94b4:       50                      push   ax
    94b5:       52                      push   dx
    94b6:       56                      push   si
    94b7:       57                      push   di
    94b8:       1e                      push   ds
    94b9:       be ef ff                mov    si,0xffef
    94bc:       bf 44 00                mov    di,0x44
    94bf:       89 fa                   mov    dx,di
    94c1:       ec                      in     al,dx
    94c2:       50                      push   ax
    94c3:       8e de                   mov    ds,si
    94c5:       ff 36 dc 01             push   WORD PTR ds:0x1dc
    94c9:       e8 1c fb                call   8fe8 <buf_add_element>
    94cc:       8e de                   mov    ds,si
    94ce:       ff 36 dc 01             push   WORD PTR ds:0x1dc
    94d2:       e8 c6 fb                call   909b <buf_count_occupied>
    94d5:       83 c4 06                add    sp,0x6
    94d8:       83 f8 32                cmp    ax,0x32
    94db:       b0 22                   mov    al,0x22
    94dd:       7e 10                   jle    94ef <ISR_10_usart0_rx+0x3b>
    94df:       50                      push   ax
    94e0:       57                      push   di
    94e1:       8e de                   mov    ds,si
    94e3:       e8 3d fc                call   9123 <usart_disable_flag>
    94e6:       83 c4 04                add    sp,0x4
    94e9:       1f                      pop    ds
    94ea:       5f                      pop    di
    94eb:       5e                      pop    si
    94ec:       5a                      pop    dx
    94ed:       58                      pop    ax
    94ee:       cf                      iret   
    94ef:       50                      push   ax
    94f0:       57                      push   di
    94f1:       8e de                   mov    ds,si
    94f3:       e8 03 fc                call   90f9 <usart_enable_flag>
    94f6:       eb ee                   jmp    94e6 <ISR_10_usart0_rx+0x32>
```

It is almost working, but only almost.

The functions called from within the ISR are accessing global data and most probably in this context the data lookup is done at wrong address.

I think the problem is here:

```
...
    94b9:       be ef ff                mov    si,0xffef   <- where it comes from?
    94bc:       bf 44 00                mov    di,0x44
    94bf:       89 fa                   mov    dx,di
    94c1:       ec                      in     al,dx
    94c2:       50                      push   ax
    94c3:       8e de                   mov    ds,si       <- why????
...
```
I am scanning through the build-ia16-master folders and looking for any explanation of applicable attributes, but I can't find anything.

That's reason why I'm stuck in trial and error loop.

PS: the interrupt vector table looks this way:
```
00000000 <ivt>:
   0:   9a 94 00 f0 9a          call   0x9af0:0x94
   5:   94                      xchg   sp,ax
   6:   00 f0                   add    al,dh
   8:   9a 94 00 f0 9a          call   0x9af0:0x94
   d:   94                      xchg   sp,ax
   e:   00 f0                   add    al,dh
  10:   9a 94 00 f0 b3          call   0xb3f0:0x94
  15:   94                      xchg   sp,ax
  16:   00 f0                   add    al,dh
  18:   b3 94                   mov    bl,0x94
  1a:   00 f0                   add    al,dh
  1c:   b3 94                   mov    bl,0x94
  1e:   00 f0                   add    al,dh
  20:   b3 94                   mov    bl,0x94
  22:   00 f0                   add    al,dh
  24:   b3 94                   mov    bl,0x94
  26:   00 f0                   add    al,dh
  28:   b4 94                   mov    ah,0x94
  2a:   00 f0                   add    al,dh
```

The code segment 0xf000: is the same for both normal and the interrupt code at least for now.

I would love to read your suggestions!

Thank you!

**Comment by ladmanj** [comment]
I have tried different memory models, but this unreasonable ds register load is still present.

I also found a discussion thread about assume and no assume ds and ss registers but even they're incompatible with interrupt or haven't any effect.

Maybe it's bug.

Is there a workaround possible, how to manually change the value which is promoted to the ds register?

I have tried to use my own inline assembly, but the value is then reverted back again immediately. 

Many thanks!

**Comment by tkchia** [comment]
Hello @ladmanj,

Currently `__attribute__ ((interrupt))` thinks that, in general, `ds` ≠ your data segment, upon function entry.  This is because &mdash; again in general &mdash; the interrupt handler can be expected to be called any time, including from third-party code.  So the compiler will output code to load `ds` with (what it thinks is) the correct segment value, for addressing the data segment.

I can probably hack `gcc-ia16` a bit to handle something like `__attribute__ ((interrupt, assume_ds_data))`.  (I am not sure I will recommend such an idiom, though.)

Alternatively...
```
    94b9:       be ef ff                mov    si,0xffef
    ...
    94c3:       8e de                   mov    ds,si
```
If `ds` = `0xffef` is not the correct value here, then in that case, what _should_ be the correct value?  How does your linker script lay out the data segment?  I believe your linker script _should_ be able to tell the linker the correct value to load into `ds` here &mdash; perhaps some simple tweaks will solve your problem.

Thank you!

**Comment by ladmanj** [comment]
Hi @tkchia 

```__attribute__ ((interrupt, assume_ds_data)) ``` makes no difference.

The correct ds value in my current case is 0x0000.
Even the whole addressable space is consisted from RAM, for sake of simplicity i have data only at zero segment (now).

I don't know what to do in the ld-script or elsewhere to tell the linker where the interrupt context data segment is or should be. But i certainly would like to.

Thank you!

**Comment by ladmanj** [comment]
I'm searching for any document regarding interaction between segment registers and ld-script but can't find anything.
Nor something to related to specific address range for data used in interrupt routines.

How would you attempt to modify that value ``` 94b9:       be ef ff                mov    si,0xffef ``` ?

How is that value determined in any case? For me it's completely unclear.

Thank you!

[project.zip](https://github.com/tkchia/gcc-ia16/files/7531874/project.zip)

**Comment by ladmanj** [comment]
Hi @tkchia 

I have no workaround, I must give it up for now.
Maybe you will come with something sooner or later.

Thank you.

**Comment by tkchia** [comment]
Hello @ladmanj,

> `__attribute__ ((interrupt, assume_ds_data))` makes no difference.

I just pushed a change (https://github.com/tkchia/gcc-ia16/commit/03cc5faa64903d07cc278cb0fadcc0fc7d620b13) which should allow `__attribute__ ((interrupt, assume_ds_data))` to work.  Please update your copy of `gcc-ia16` and let me know if it works now.

Thank you!

**Comment by ladmanj** [comment]
Unfortunately there is some problem

```
make
ia16-elf-gcc -mcmodel=small -Werror -Os -ffunction-sections -fno-stack-protector -Iarch_8086 -I/home/ladmanj/workspace/rwt_memtest -march=i8088 -mno-segelf -o main.o -c main.c
ia16-elf-gcc -mcmodel=small -Werror -Os -ffunction-sections -fno-stack-protector -Iarch_8086 -I/home/ladmanj/workspace/rwt_memtest -march=i8088 -mno-segelf -o memtest.o -c memtest.c
ia16-elf-gcc -mcmodel=small -Werror -Os -ffunction-sections -fno-stack-protector -Iarch_8086 -I/home/ladmanj/workspace/rwt_memtest -march=i8088 -mno-segelf -o print.o -c print.c
ia16-elf-gcc -mcmodel=small -Werror -Os -ffunction-sections -fno-stack-protector -Iarch_8086 -I/home/ladmanj/workspace/rwt_memtest -march=i8088 -mno-segelf -o gdbstub.o -c gdbstub.c
ia16-elf-gcc -mcmodel=small -Werror -Os -ffunction-sections -fno-stack-protector -Iarch_8086 -I/home/ladmanj/workspace/rwt_memtest -march=i8088 -mno-segelf -o buf/buf_buffer.o -c buf/buf_buffer.c
ia16-elf-gcc -mcmodel=small -Werror -Os -ffunction-sections -fno-stack-protector -Iarch_8086 -I/home/ladmanj/workspace/rwt_memtest -march=i8088 -mno-segelf -o arch_8086/io.o -c arch_8086/io.c
arch_8086/io.c: In function 'ISR_10_usart0_rx':
arch_8086/io.c:301:1: internal compiler error: Segmentation fault
 }
 ^
unrecognized DWARF version in .debug_info at 6
Please submit a full bug report,
with preprocessed source if appropriate.
See <http://gcc.gnu.org/bugs.html> for instructions.
make: *** [Makefile:86: arch_8086/io.o] Error 1
rm buf/buf_buffer.o gdbstub.o memtest.o print.o main.o
```

Before this bug showed (before debug print statements were cleaned off), the code was built once , the functions were accessing its data, but the assembly listing wasn't looking near optimal.

[project.zip](https://github.com/tkchia/gcc-ia16/files/7532963/project.zip)

**Comment by tkchia** [comment]
Hello @ladmanj,

There was a bug in the compiler.  I have just pushed out another patch (https://github.com/tkchia/gcc-ia16/commit/e76c4efcde8e430654228753d9bc77787f1de4ed).  Thank you!

**Comment by ladmanj** [comment]
I will try it a little bit later and then let you know.

Thank you.

**Comment by ladmanj** [comment]
Hi @tkchia 

I'm sorry for the delay.

The compiler no longer crashes, also the code is running ok.
I have problems with optimization though, because i am coding interaction between ISR and normal code, I'm using volatile variables and interrupt disabling and enabling again, but even volatile, some are seemingly optimized out when compiled without printing-out.

Maybe it's not the compiler fault but mine, also the hardware behaves a little bit unexpectedly so honestly I'm too confused to make any statement.

I think this thread should be closed and maybe I'll open a different one as the situation gets clearer.

Thank you

**Comment by wickerwaka** [comment]
I was in a similar situation and found this thread so I wanted to add my own solution in case anyone else comes across it. I'm working with a ROM based system. ROM is mapped from 0x00000 - 0x7ffff and RAM is mapped from 0xe0000 - 0xeffff. I'm using the "Small" memory model, so `CS` to set to `0x0000` and `DS`, `ES`, `SS` are set to `0xe000` in my entry function. 

In my linker script I was doing this: 
```
    .text 0x00000 : {
        KEEP(*(.vectors .vectors.*));
        . = ALIGN(4);
        *(.text);
        *(.text.*);
        *(.entry.text);
        . = ALIGN(16);
    }

    . = ALIGN(16);

    data_load = . ;

    .data 0x0000 : AT(data_load) {
        data_start = . ;
        *(.rodata .rodata.*);
        *(.data .data.*);
        data_end = . ;

        . = ALIGN(4);
        bss_start = . ;
        *(.bss .bss.*);
        *(COMMON);
        bss_end = . ;
        . = ALIGN(16);
    }
```
So the data section was placed after the text section in the rom file (`data_load` address) and then in my entry function I was copying the data from `data_load` into the data segment `0xe000`.

The problem was the linker was determine the default data segment from the LMA of the data section, which was `data_load` in this case which is its offset in the ROM image. Changing the data section definition to:
```.data 0x0000 : AT(0xe0000)```
Fixed this, now the linker assumed the default data segment is `0xe000` and it correctly sets the `DS` register inside of interrupt handlers.

However now the offset of the data section in ROM file generated by `objcopy` was at 0xe0000, to fix this I remapped the section address during the objdump step like so:
```objcopy -O binary --change-section-lma .data=0x10000 main.elf main.bin```
The ROM now consists of 64kb of code followed by upto 64kb of data and I changed my initialization code to copy any data from segment 0x1000 to 0xe000.
